### PR TITLE
Get the max session len from config.json

### PR DESCRIPTION
**Pull request by AllentDan**

### CODE REVIEW

**Comment by lvhan028** [comment]
May add UT to verify the session_len for a bunch of supported models, including VLMs

**Comment by AllentDan** [comment]
> May add UT to verify the session_len for a bunch of supported models, including VLMs

It seems we can only add a UT for `_get_and_verify_max_len` func which is copied from vllm.

File: lmdeploy/utils.py [hunk 1]
```diff
@@ -255,3 +256,77 @@ async def __tmp():
...
+                else max_len_key
+            derived_max_model_len = min(derived_max_model_len, max_len)
+    if derived_max_model_len == float('inf'):
+        if max_model_len is not None:
+            # If max_model_len is specified, we use it.
+            return max_model_len
+
+        default_max_len = 2048
+        logger = get_logger('lmdeploy')
+        logger.warning(
```
Comment (lvhan028) [inline lmdeploy/utils.py:299]: logger.warning(f'')

File: lmdeploy/utils.py [hunk 2]
```diff
@@ -255,3 +256,77 @@ async def __tmp():
...
+        max_model_len = int(derived_max_model_len)
+    elif max_model_len > derived_max_model_len:
+        # Some models might have a separate key for specifying model_max_length
+        # that will be bigger than derived_max_model_len. We compare user input
+        # with model_max_length and allow this override when it's smaller.
+        model_max_length = getattr(hf_tm_config, 'model_max_length', None)
+        if model_max_length is not None and max_model_len <= model_max_length:
+            pass
+        else:
+            raise ValueError(
```
Comment (lvhan028) [inline lmdeploy/utils.py:325]: The `session_len` of InternLM2 can be 200k, which is much greater than `derived_max_model_len`

Comment (AllentDan) [reply lmdeploy/utils.py:325]: shall we just give a warning here?

File: lmdeploy/utils.py [hunk 3]
```diff
@@ -255,3 +256,77 @@ async def __tmp():
...
+        derived_max_model_len = default_max_len
+
+    rope_scaling = getattr(hf_tm_config, 'rope_scaling', None)
+    if rope_scaling is not None and rope_scaling['type'] != 'su':
+        assert 'factor' in rope_scaling
+        scaling_factor = rope_scaling['factor']
+        if rope_scaling['type'] == 'yarn':
+            derived_max_model_len = rope_scaling[
+                'original_max_position_embeddings']
+        derived_max_model_len *= scaling_factor
```
Comment (lvhan028) [inline lmdeploy/utils.py:313]: @lzhangzz does it make sense?

**Review by lvhan028 — DISMISSED** [review]

**Comment by lvhan028** [comment]
The converter should be updated too

**Comment by lvhan028** [comment]
converter uses sesssion_len from the chat template. I think we can change it to `_get_and_verify_max_len`

**Comment by AllentDan** [comment]
converter now uses `max_position_embeddings`. May check *lmdeploy/turbomind/deploy/target_model/base.py*

**Comment by lvhan028** [comment]
Do we need to remove `session_len` from chat template?

**Comment by AllentDan** [comment]
> Do we need to remove `session_len` from chat template?

triton server uses it.

**Review by lvhan028 — DISMISSED** [review]

File: lmdeploy/utils.py [hunk 4]
```diff
@@ -255,3 +256,77 @@ async def __tmp():
...
+        # with model_max_length and allow this override when it's smaller.
+        model_max_length = getattr(hf_tm_config, 'model_max_length', None)
+        if model_max_length is not None and max_model_len <= model_max_length:
+            pass
+        else:
+            logger.warning(
+                f'User-specified max_model_len ({max_model_len}) is greater '
+                'than the derived max_model_len '
+                f'({max_len_key}={derived_max_model_len} or model_max_length='
+                f"{model_max_length} in model's config.json). This may lead "
```
Comment (lvhan028) [inline lmdeploy/utils.py:329]: 把 "This may lead" 开始的日志去掉吧。主要是不想在 internlm2 200k 上下文的时候，引起不必要的误会

**Review by lvhan028 — APPROVED** [review]

**Review by irexyc — APPROVED** [review]

**Pull request merged**